src/token-providing-subscription.ts

```diff
@@ -1,147 +1,180 @@
-import { ErrorResponse, ElementsHeaders } from './network';
+import { Logger } from './logger';
+import { ElementsHeaders, ErrorResponse } from './network';
+import {
+  SubscribeStrategy,
+  SubscribeStrategyListeners,
+} from './subscribe-strategy';
 import {
-    Subscription,
-    SubscriptionConstructor,
-    SubscriptionEvent,
-    SubscriptionState,
+  Subscription,
+  SubscriptionConstructor,
+  SubscriptionEvent,
+  SubscriptionState,
 } from './subscription';
-import { Logger } from './logger';
 import { TokenProvider } from './token-provider';
-import { SubscribeStrategy, SubscribeStrategyListeners } from './subscribe-strategy';
-import { PCancelable } from 'p-cancelable';
-
-export let createTokenProvidingStrategy: (tokenProvider: TokenProvider, nextSubscribeStrategy: SubscribeStrategy, logger: Logger) => SubscribeStrategy = 
-(tokenProvider, nextSubscribeStrategy, logger) => {
-
-    class TokenProvidingSubscription implements Subscription {
-
-        private state: SubscriptionState;
 
-        private onTransition = (newState: SubscriptionState) => {
-            this.state = newState;    
+import PCancelable = require('p-cancelable');
+
+export let createTokenProvidingStrategy: (
+  nextSubscribeStrategy: SubscribeStrategy,
+  logger: Logger,
+  tokenProvider?: TokenProvider,
+) => SubscribeStrategy = (nextSubscribeStrategy, logger, tokenProvider) => {
+  // Token provider might not be there. If missing, go straight to the underlying subscribe strategy
+  if (tokenProvider) {
+    return (listeners, headers) =>
+      new TokenProvidingSubscription(
+        tokenProvider,
+        logger,
+        nextSubscribeStrategy,
+        listeners,
+        headers,
+      );
+  } else {
+    return (listeners, headers) => nextSubscribeStrategy(listeners, headers);
+  }
+};
+
+class TokenProvidingSubscription implements Subscription {
+  private state: SubscriptionState;
+
+  constructor(
+    tokenProvider: TokenProvider,
+    logger: Logger,
+    nextSubscribeStrategy: SubscribeStrategy,
+    listeners: SubscribeStrategyListeners,
+    headers: ElementsHeaders,
+  ) {
+    class TokenProvidingState implements SubscriptionState {
+      private underlyingSubscription: Subscription;
+      private tokenPromise: PCancelable.PCancelable<string>;
+
+      constructor(private onTransition: (newState: SubscriptionState) => void) {
+        logger.verbose(
+          `TokenProvidingSubscription: transitioning to TokenProvidingState`,
+        );
+
+        const isTokenExpiredError: (error: any) => boolean = error => {
+          return (
+            error instanceof ErrorResponse &&
+            error.statusCode === 401 &&
+            error.info === 'authentication/expired'
+          );
+        };
+
+        const fetchTokenAndExecuteSubscription = () => {
+          this.tokenPromise = new PCancelable<string>(
```

Comment:
I'm not sure what the expectation is here in terms of whether or not the tokenProvider `fetchToken` call should be cancelable or if you want to be able to cancel the attempt to establish a subscription, or something else entirely.